tests/elf_file.py

```diff
@@ -0,0 +1,217 @@
+
+        @dataclass
+        class Section:
+            sh_name: int
+            sh_type: int
+            sh_flags: int
+            sh_addr: int
+            sh_offset: int
+            sh_size: int
+            sh_link: int
+            sh_info: int
+            sh_addralign: int
+            sh_entsize: int
+
+        sections = [
+            Section(*read("IIPPPPIIPP", e_shoff + i * e_shentsize))
+            for i in range(e_shnum)
+        ]
+        shstrtab_offset = sections[e_shstrndx].sh_offset
+        sec_names = [read_string(shstrtab_offset, s.sh_name) for s in sections]
+
+        strtab_sections = [
+            i
+            for i in range(e_shnum)
+            if sections[i].sh_type == SHT_STRTAB and i != e_shstrndx
+        ]
+        assert len(strtab_sections) == 1
+        strtab_offset = sections[strtab_sections[0]].sh_offset
+
+        symtab_sections = [
+            i for i in range(e_shnum) if sections[i].sh_type == SHT_SYMTAB
+        ]
+        assert len(symtab_sections) == 1
+        symtab = sections[symtab_sections[0]]
+
+        elf = ElfFile()
+
+        # Parse SHT_PROGBIT/SHT_NOBITS sections (.text, .data, .bss, etc.)
+        for s, name in zip(sections, sec_names):
+            if s.sh_type == SHT_PROGBITS:
+                if name == ".comment" or name.startswith(".note"):
+                    continue
+                section_data = data[s.sh_offset : s.sh_offset + s.sh_size]
+                elf.sections[name] = ElfSection(
+                    address=s.sh_addr, name=name, data=section_data
+                )
+            elif s.sh_type == SHT_NOBITS and s.sh_size != 0:
+                section_data = b"\0" * s.sh_size
+                elf.sections[name] = ElfSection(
+                    address=s.sh_addr, name=name, data=section_data
+                )
+
+        # Parse SHT_SYMTAB section (symbol table)
+        symbols_by_index: List[ElfSymbol] = []
+        for offset in range(0, symtab.sh_size, symtab.sh_entsize):
+            if is_32bit:
+                sym_offset = symtab.sh_offset + offset
+                st_name, st_value, st_size, st_info, st_other, st_shndx = read(
+                    "IIIBBH", sym_offset
+                )
+            else:
+                sym_offset = symtab.sh_offset + offset
+                st_name, st_info, st_other, st_shndx, st_value, st_size = read(
+                    "IBBHQQ", sym_offset
+                )
+            sym_name = read_string(strtab_offset, st_name)
+            if st_shndx == 0 or st_shndx >= 0xFF00:
+                section = None
+            else:
+                section = elf.sections.get(sec_names[st_shndx])
+
+            symbol = ElfSymbol(offset=st_value, name=sym_name, section=section)
+            elf.symbols[sym_name] = symbol
+            symbols_by_index.append(symbol)
+            if section is not None and (
+                st_value not in section.symbols or not sym_name.startswith("...")
```

Comment:
Added the following comment:
```
            # Do not overwrite existing symbols at this address, unless the existing
            # symbol starts with "..." (such as `...bss@0`).
```

I was seeing `...bss@0` as a symbol pointing to the start of the `.bss` section. This would shadow the actual symbol that we cared about at `.bss+0`. (The ordering of these two in the symtab was inconsistent.)

There may be other symbols we'd want to ignore? Also `mips_to_c` was not able to parse the symbol name `...bss@0`, but I'm not sure if that needs to be fixed...

Reply:
Technically that comment doesn't exactly match what the code does (it allows overwriting a non-`...` symbol with another one), but I see what you mean!